optaplanner-core/src/test/java/org/optaplanner/core/config/solver/EnvironmentModeTest.java

```diff
@@ -0,0 +1,237 @@
+
+        solverConfig = new SolverConfig()
+                .withSolutionClass(TestdataSolution.class)
+                .withEntityClasses(TestdataEntity.class)
+                .withEnvironmentMode(environmentMode)
+                .withPhases(initializerPhaseConfig, localSearchPhaseConfig);
+    }
+
+    @Parameterized.Parameters(name = "{0}")
+    public static Collection<EnvironmentMode> params() {
+        return Arrays.asList(EnvironmentMode.values());
+    }
+
+    @Test
+    public void determinismTest() {
+        setSolverConfigCalculatorClass(TestdataDifferentValuesCalculator.class);
+
+        Solver solver1 = SolverFactory.create(solverConfig).buildSolver();
+        Solver solver2 = SolverFactory.create(solverConfig).buildSolver();
+
+        switch (environmentMode) {
+            case PRODUCTION:
+            case NON_REPRODUCIBLE:
+                assertNonReproducibility(solver1, solver2);
+                break;
+            case FULL_ASSERT:
+            case FAST_ASSERT:
+            case NON_INTRUSIVE_FULL_ASSERT:
+            case REPRODUCIBLE:
+                assertReproducibility(solver1, solver2);
+                break;
+            default:
+                throw new AssertionError("Missing case for " + environmentMode);
+        }
+    }
+
+    @Test
+    public void corruptedCustomMovesTest() {
+        // intrusive modes should throw exception about corrupted undoMove
+        setSolverConfigCalculatorClass(TestdataDifferentValuesCalculator.class);
+
+        switch (environmentMode) {
+            case FULL_ASSERT:
+            case FAST_ASSERT:
+                setSolverConfigMoveListFactoryClassToCorrupted(TestdataChangeMoveWithCorruptedUndoMoveFactory.class);
+                assertIllegalStateExceptionWhileSolving("corrupted undoMove");
+                break;
+            case NON_INTRUSIVE_FULL_ASSERT:
+                setSolverConfigMoveListFactoryClassToCorrupted(TestdataCorruptedEntityUndoMoveFactory.class);
+                assertIllegalStateExceptionWhileSolving("not the uncorruptedScore");
+                break;
+            case REPRODUCIBLE:
+            case NON_REPRODUCIBLE:
+            case PRODUCTION:
+                // no exception expected
+                break;
+            default:
+                throw new AssertionError("Missing case for " + environmentMode);
+        }
+    }
+
+    @Test
+    public void corruptedScoreRulesTest() {
+        // for full assert modes it should throw exception about corrupted score
+        setSolverConfigCalculatorClass(TestdataCorruptedDifferentValuesCalculator.class);
+
+        switch (environmentMode) {
+            case FULL_ASSERT:
+            case NON_INTRUSIVE_FULL_ASSERT:
+                assertIllegalStateExceptionWhileSolving("not the uncorruptedScore");
+                break;
+            case FAST_ASSERT:
+                assertIllegalStateExceptionWhileSolving("Score corruption analysis could not be generated ");
```

Comment:
There is a weird inconsistency. Sometimes you start the `String` with a capital letter, sometimes not.

Reply:
The String is a part of an exception message (not the whole message), sometimes it's mid-sentence and needs to be lower-case.